src/gnc_attitude_estimator.cpp

```diff
@@ -55,26 +387,92 @@ AttitudeEstimate::AttitudeEstimate()
   is_valid(false) { }
 
 void attitude_estimator_reset(AttitudeEstimatorState &state,
-    double t, lin::Vector3d const &r_ecef, lin::Vector3f const &b_body,
-    lin::Vector3f const &s_body) {
-  GNC_ASSERT_NORMALIZED(s_body);
+    double t, lin::Vector4f const &q_body_eci) {
+  GNC_ASSERT_NORMALIZED(q_body_eci);
 
-  // TODO : Implement this
-  state = AttitudeEstimatorState();
+  /* Default, initial attitude covariance (units of radiancs squared). */
+  GNC_TRACKED_CONSTANT(constexpr static float, var_q, 0.0305);
+  /* Default, initial gyro bias covariance (units of radians per second all
+   * squared). */
+  GNC_TRACKED_CONSTANT(constexpr static float, var_g, 0.0049);
+  /* Default, initial state. */
+  GNC_TRACKED_CONSTANT(constexpr static UkfVector6, init_state, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0);
+
+  // Set the proper fields in state
+  state.t = t;
+  state.q = q_body_eci;
+  state.x = init_state;
+  state.P = lin::zeros<decltype(state.P)>();
+  state.P(0, 0) = var_q;
+  state.P(1, 1) = state.P(0, 0);
+  state.P(2, 2) = state.P(0, 0);
+  state.P(3, 3) = var_g;
+  state.P(4, 4) = state.P(3, 3);
+  state.P(5, 5) = state.P(4, 4);
+  state.is_valid = true;
+
+  // If invalid, set everything back to NaNs
+  if (!lin::isfinite(state.t) ||
+      !lin::all(lin::isfinite(state.q)) ||
+      !lin::all(lin::isfinite(state.x)) ||
+      !lin::all(lin::isfinite(state.P)))
+    state = AttitudeEstimatorState();
 }
 
 void attitude_estimator_reset(AttitudeEstimatorState &state,
-    double t, lin::Vector4f const &q_eci_body) {
-  GNC_ASSERT_NORMALIZED(q_eci_body);
+    double t, lin::Vector3d const &r_ecef, lin::Vector3f const &b_body,
+    lin::Vector3f const &s_body) {
+  GNC_ASSERT_NORMALIZED(s_body);
 
   // TODO : Implement this
-  state = AttitudeEstimatorState();
+  // https://github.com/pathfinder-for-autonomous-navigation/psim/issues/187
+  lin::Vector4f q_eci_body = lin::nans<float, 4, 1>();
+  attitude_estimator_reset(state, t, q_eci_body);
 }
 
 void attitude_estimator_update(AttitudeEstimatorState &state,
     AttitudeEstimatorData const &data, AttitudeEstimate &estimate) {
-  // TODO : Implement this
-  state = AttitudeEstimatorState();
-  estimate = AttitudeEstimate();
+  // Ensure we have a valid state and all the required inputs
+  if (!state.is_valid ||
+      !lin::all(lin::isfinite(data.r_ecef)) ||
+      !lin::all(lin::isfinite(data.b_body)) ||
+      !lin::all(lin::isfinite(data.w_body)) ||
+      !lin::isfinite(data.t)) {
+    // Set the state and estimate to be invalid
+    state = AttitudeEstimatorState();
```

Comment:
Can the filter just propagate without sensor readings? Or does the state needs to be reset.

Reply:
You can't just propagate without sensor readings, at least that's how it's coded now. In terms of our use case however, there won't be a cc where we don't have at least magnetometer and gyroscope data. If for some reason we don't have that information, we have bigger problems than our attitude filter too lol